Compression/src/CompressDetStepMCs_module.cc

```diff
@@ -615,17 +618,15 @@ void mu2e::CompressDetStepMCs::updateStepPointMCs() {
 }
 
 void mu2e::CompressDetStepMCs::updateMCTrajectories() {
-  for (auto& i_mcTrajectory : *_newMCTrajectories) {
-    const auto& oldSimPtr = i_mcTrajectory.second.sim();
+  for (auto& i_mcTrajectory : _newMCTrajs) {
+    const auto& oldSimPtr = i_mcTrajectory.sim();
     art::Ptr<mu2e::SimParticle> newSimPtr = _simPtrRemap.at(oldSimPtr);
     if(_debugLevel>0) {
       std::cout << "Updating SimParticlePtr in MCTrajectory from " << oldSimPtr << " to " << newSimPtr << std::endl;
     }
-    i_mcTrajectory.second.sim() = newSimPtr;
-    auto mcTrajPair = _newMCTrajectories->extract(i_mcTrajectory.first);
-    mcTrajPair.key() = newSimPtr;
-    _newMCTrajectories->insert(std::move(mcTrajPair));
-    //    i_mcTrajectory.first = newSimPtr;
+    // overwrite the internal ptr and insert with the new key
+    i_mcTrajectory.sim() = newSimPtr;
+    _newMCTrajectories->insert(std::make_pair(i_mcTrajectory.sim(), i_mcTrajectory));
   }
```

Comment:
I believe that at the end of this loop your are finished with _newMCTrajs for this event.   I suggest that you clear the array here, instead of at the start of the next event.   If you need to keep it around for debugging, clear it immediately before returning from produce().

If time is tight, feel free to make a separate PR.

Reply:
Currently the clear() occurs at the same place _newMCTrajectories is created, underlining their connection in a direct way.  What risk or cost is there to have the content persist until the next event?

Reply:
The cost is that it increases the VmHWM because the memory is retained throughout the remainder of event, even though it no longer has any value.  While I do not have a quantitative feel for the cost in this particular case, it's something we don't want to establish as a recommended pattern.

At one time LArSoft was carrying GBs around this way.  According to the Paterno et al, the better solution is to make _newMCTrajs a local variable that is instantiated at the start of produce and pass it  by ref or const-ref into the member functions that use it.  That way the memory is automatically released when it goes out of scope at the end of produce().  But that's a more intrusive change.

Reply:
OK that seems legit.

Reply:
Thanks.